File: api/auth_api.py
"""
Модуль: auth_api.py
Назначение: REST API аутентификации и авторизации
Автор: Разработчик
Дата создания: 21.03.2026
Требования: LLR_SecurityManager_Authenticate_01, Functional.Security.UserManagement
"""

from flask import Blueprint, jsonify, request
from flask_login import current_user, login_required, login_user, logout_user

from services.audit_service import log_action
from services.user_service import UserError, authenticate_user, register_user

auth_api = Blueprint("auth_api", __name__, url_prefix="/api/auth")


@auth_api.route("/login", methods=["POST"])
def login():
    """
    Назначение: Аутентификация пользователя
    Id требования: LLR_SecurityManager_Authenticate_01
    Входные данные: JSON {username, password}
    Выходные данные: JSON с данными пользователя
    """
    data = request.get_json(silent=True) or {}
    username = data.get("username", "").strip()
    password = data.get("password", "")

    if not username or not password:
        return jsonify({"error": "Имя пользователя и пароль обязательны"}), 400

    user = authenticate_user(username, password)
    if user is None:
        return jsonify({"error": "Неверное имя пользователя или пароль"}), 401

    login_user(user)
    return jsonify({
        "id": user.id,
        "username": user.username,
        "email": user.email,
        "role": user.role,
    })


@auth_api.route("/register", methods=["POST"])
def register():
    """
    Назначение: Регистрация нового пользователя
    Id требования: Functional.Security.UserManagement
    Входные данные: JSON {username, email, password, role}
    Выходные данные: JSON с данными пользователя
    """
    data = request.get_json(silent=True) or {}
    try:
        user = register_user(
            username=data.get("username", "").strip(),
            email=data.get("email", "").strip(),
            password=data.get("password", ""),
            role=data.get("role", "analyst"),
        )
        login_user(user)
        return jsonify({
            "id": user.id,
            "username": user.username,
            "email": user.email,
            "role": user.role,
        }), 201
    except UserError as exc:
        return jsonify({"error": exc.message, "code": exc.code}), 400


@auth_api.route("/logout", methods=["POST"])
@login_required
def logout():
    """Назначение: Выход из системы."""
    logout_user()
    return jsonify({"message": "OK"})


@auth_api.route("/me", methods=["GET"])
@login_required
def me():
    """Назначение: Получение данных текущего пользователя."""
    return jsonify({
        "id": current_user.id,
        "username": current_user.username,
        "email": current_user.email,
        "role": current_user.role,
    })

File: backend/src/api/auth_api.cpp
/*
Модуль: auth_api.cpp
Назначение: Реализация REST API аутентификации
Автор: Разработчик
Дата создания: 21.03.2026
Требования: LLR_SecurityManager_Authenticate_01
*/

#include "auth_api.hpp"
#include "middleware.hpp"
#include "session_manager.hpp"
#include "services/user_service.hpp"

void register_auth_routes(httplib::Server& svr)
{
    svr.Post("/api/auth/login", [](const httplib::Request& req,
        httplib::Response& res)
    {
        auto body = parse_json_body(req);
        std::string username = body.value("username", "");
        std::string password = body.value("password", "");

        if (username.empty() || password.empty()) {
            json_error(res, "Имя пользователя и пароль обязательны", 400);
            return;
        }

        auto user = authenticate_user(username, password);
        if (!user.has_value()) {
            json_error(res, "Неверное имя пользователя или пароль", 401);
            return;
        }

        auto token = SessionManager::instance().create_session(user->id);
        res.set_header("Set-Cookie",
            "session=" + token + "; Path=/; HttpOnly; SameSite=Lax");
        json_response(res, {
            {"id", user->id}, {"username", user->username},
            {"email", user->email}, {"role", user->role}
        });
    });

    svr.Post("/api/auth/register", [](const httplib::Request& req,
        httplib::Response& res)
    {
        auto body = parse_json_body(req);
        try {
            auto user = register_user(
                body.value("username", ""),
                body.value("email", ""),
                body.value("password", ""),
                body.value("role", "analyst"));
            auto token = SessionManager::instance().create_session(user.id);
            res.set_header("Set-Cookie",
                "session=" + token + "; Path=/; HttpOnly; SameSite=Lax");
            json_response(res, {
                {"id", user.id}, {"username", user.username},
                {"email", user.email}, {"role", user.role}
            }, 201);
        } catch (const UserError& e) {
            json_error(res, e.what(), 400);
        }
    });

    svr.Post("/api/auth/logout", [](const httplib::Request& req,
        httplib::Response& res)
    {
        auto cookie_it = req.headers.find("Cookie");
        if (cookie_it != req.headers.end()) {
            auto& sm = SessionManager::instance();
            std::string token = sm.extract_token(cookie_it->second);
            sm.destroy_session(token);
        }
        res.set_header("Set-Cookie",
            "session=; Path=/; HttpOnly; Max-Age=0");
        json_response(res, {{"message", "OK"}});
    });

    svr.Get("/api/auth/me", [](const httplib::Request& req,
        httplib::Response& res)
    {
        auto user = require_auth(req, res);
        if (!user) return;
        json_response(res, {
            {"id", user->id}, {"username", user->username},
            {"email", user->email}, {"role", user->role}
        });
    });
}
